[lldb] Add some vector operations to the IRInterpreter

dsandersllvm opened the pull request:
This allows the debugger to evaluate expressions without requiring the
expression to be CodeGen'd and executed on the target. This should be more
efficient for many existing targets but is necessary for targets which are
not yet able to evaluate on the target.

As far as I know most targets have a vector memory layout that matches the
IR element order. Most little endian targets choose to use a little endian
element order, and two out of the three big endian targets I know of
(MIPS MSA and ARM NEON) choose to use little endian element order even
when the elements are big endian which matches LLVM-IR's order. The third
is PowerPC Altivec which has the highest indexed element first for
big-endian mode.

I've attempted to implement the correct element ordering on the relevant
operations but I don't really have a means to test the case where the
element order doesn't match LLVM-IR's element order so I've chosen to have
a guard against element order mismatches to ensure that this change can't
break expression evaluation on those targets.

dsandersllvm committed:
[lldb] Add some vector operations to the IRInterpreter

This allows the debugger to evaluate expressions without requiring the
expression to be CodeGen'd and executed on the target. This should be more
efficient for many existing targets but is necessary for targets which are
not yet able to evaluate on the target.

In terms of memory layout, we have:
           | Element Values   | Element Order |
  ARM NEON | Endian-dependant | Zero-first    |
  MIPS MSA | Endian-dependant | Zero-first    |
  PowerPC  | Endian-dependant | Zero-first    |
  SystemZ  | Endian-dependant | Zero-first    |
Where Zero-first means that element zero of an array/vector is at the lowest
memory address.

In terms of register layout things, we have:
           | Element Values   | Element Order     | Effective Element Order
  ARM NEON | Endian-dependant | Zero-first        | Zero-first
  MIPS MSA | Endian-dependant | Zero-first        | Zero-first
  PowerPC  | Endian-dependant | Endian-dependant* | Zero-first (lane-swaps LE)
  SystemZ  | Endian-dependant | Endian-dependant* | Zero-first (lane-swaps BE)
*HW is endian-dependent but CodeGen accounts for it

PowerPC is a little more complicated than shown above as it actually supports
two modes: True-LE and Big-on-Little and the above table shows True-LE's
behaviour. See
https://llvm.org/devmtg/2014-10/Slides/Schmidt-SupportingVectorProgramming.pdf
I haven't seen evidence that big-on-little is implemented yet so I haven't
attempted to account for it.

The end result of this is that transferring values between llvm-ir and memory
is consistent between the four targets but transferring values between llvm-ir
and registers potentially requires transformations. I've therefore:
* Redefined GetVectorElementOrder to refer to the register layout not memory
* Made Materializer/Dematerializer bail out when reading/writing vector values.
* Made bitcast bail out for the cases where a bitcast is a shuffle rather than
  a nop.

Michael137 commented on lldb/source/Plugins/Architecture/PPC64/ArchitecturePPC64.h:
```diff
@@ -30,9 +30,14 @@ class ArchitecturePPC64 : public Architecture {
 private:
   static std::unique_ptr<Architecture> Create(const ArchSpec &arch);
-  ArchitecturePPC64() = default;
+  ArchitecturePPC64(lldb::ByteOrder vector_element_order)
```
Can we split the architecture plugin changes into a separate PR?

dsandersllvm replied:
It's not the only big-endian target but it is the only target I know of where the order of vector elements doesn't match LLVM-IR's order. MIPS  and ARM both have big-endian modes but vectors are 0-element first in both endians whereas big-endian PowerPC is highest-indexed element first. If I hadn't handled this then we'd read/write their vectors in reversed element order every time we tried to copy memory to/from an LLVM-IR value.

MIPS and ARM's vector layout has a different quirk on LLVM-IR/memory which is that bitcast isn't a no-op, it's a shuffle (which bytes swaps depends on the types involved). This is because it's defined as a store of the original type followed by a load of the new type. I haven't implemented this yet because I didn't need to support vector bitcast.

Michael137 commented on lldb/source/Expression/IRInterpreter.cpp:
```diff
@@ -1564,7 +1660,103 @@ bool IRInterpreter::Interpret(llvm::Module &module, llvm::Function &function,
+      }
+
+      if (log) {
+        LLDB_LOGF(log, "Interpreted an ExtractElement");
+        LLDB_LOGF(log, "  Vector: 0x%" PRIx64, vector_addr);
+        LLDB_LOGF(log, "  Index: %llu", (unsigned long long)index);
+        LLDB_LOGF(log, "  Element offset: %zu", element_offset);
+        LLDB_LOGF(log, "  Result: 0x%" PRIx64, result_addr);
```
Can you combin these into a single `LLDB_LOG`? You can replace the printf specificers with `{0}, {1}, etc.` too

Michael137 reviewed (commented):
For testing the endianess, I'm not sure we have PPC LLDB buildbots that would test this (@DavidSpickett or @JDevlieghere might know). Any chance this can be tested with unit-tests somehow? Maybe we can just pass an LLVM IR module into the IRInterpreter? We could potentially build something into `lldb-test` even. If that's too much hassle we can also just do that as a follow-up

dsandersllvm commented:
> For testing the endianess, I'm not sure we have PPC LLDB buildbots that would test this (@DavidSpickett or @JDevlieghere might know). Any chance this can be tested with unit-tests somehow? Maybe we can just pass an LLVM IR module into the IRInterpreter? We could potentially build something into `lldb-test` even. If that's too much hassle we can also just do that as a follow-up

I'm not sure we can. The problem arises from transferring LLVM-IR values to/from the targets memory. If we don't have the target consuming/producing that memory then our test could just be consistently-wrong and pass anyway.

dsandersllvm committed:
[lldb] Architecture plugins should report the vector element order. NFC

Some targets like PowerPC store their
- PowerPC little endian:  little endian elements ordered 0, 1, 2, ...
- PowerPC big endian:     big endian elements ordered n-1, n-2, n-3, ...
- ARM/MIPS little endian: little endian elements ordered 0, 1, 2, ...
- ARM/MIPS big endian:    big endian elements ordered 0, 1, 2, ...

This matters when LLVM-IR values are transferred to/from target memory since
LLVM-IR orders elements 0, 1, 2, ... regardless of endianness.

This will be used in #155000 by changes to the IRInterpreter to allow it to
evaluate some vectors without executing on the target

DavidSpickett commented:
> For testing the endianess, I'm not sure we have PPC LLDB buildbots that would test this

Debian will build PPC little endian once in a while and somewhere in the depths of IBM there might be a machine but I've never seen a bug report from one.

Also not sure what endians they ever supported. ISTR Linux is only supported on LE now but AIX is BE only.

dsandersllvm committed:
[lldb] Add type hints to gdbclientutils.py and fix issues

Everything in this should be python 3.9. The docs say the minimum is 3.8
but there's existing code in this suite that needs 3.9 so I think 3.9 is
ok.

Issues:
qEcho() is passed an argument by the callers that the function didn't have
Several functions in the base class would silently do nothing if not
overriden. These now use @abstractmethod to require overrides
sendall() had inconsistent return types between overrides

dsandersllvm committed:
[lldb] Add bidirectional packetLog to gdbclientutils.py

While debugging the tests for PR155000 I found it helpful to have both sides
of the simulated gdb-rsp traffic rather than just the responses so I've added
a packetLog to MockGDBServer. The existing response-only one in
MockGDBServerResponder is used by tests so I chose not to change it
